src/lib.rs

```diff
@@ -69,75 +110,87 @@ pub unsafe extern "C" fn pthread_join(
     native: libc::pthread_t,
     _value: *mut *mut libc::c_void,
 ) -> libc::c_int {
-    if native == MAIN_THREAD_ID {
-        // This is not a valid thread to join on
-        return libc::EINVAL;
-    }
+    let pthread = Box::from_raw(native as *mut PThread);
 
-    let result = ctru_sys::threadJoin(native as ctru_sys::Thread, u64::MAX);
+    let result = ctru_sys::threadJoin(pthread.thread, u64::MAX);
     if ctru_sys::R_FAILED(result) {
         // TODO: improve the error code by checking the result further?
         return libc::EINVAL;
     }
 
-    ctru_sys::threadFree(native as ctru_sys::Thread);
+    ctru_sys::threadFree(pthread.thread);
 
     0
 }
 
 #[no_mangle]
-pub unsafe extern "C" fn pthread_detach(thread: libc::pthread_t) -> libc::c_int {
-    if thread == MAIN_THREAD_ID {
-        // This is not a valid thread to detach
-        return libc::EINVAL;
-    }
-
-    ctru_sys::threadDetach(thread as ctru_sys::Thread);
+pub unsafe extern "C" fn pthread_detach(native: libc::pthread_t) -> libc::c_int {
+    let pthread = Box::from_raw(native as *mut PThread);
```

Comment:
Hmm, in that case how would the `PThread` ever get freed? We'd probably need to set up a thread local destructor for detached threads to clean themselves up when finished, which ties in the whole discussion around https://github.com/Meziu/rust-horizon/pull/16

Reply:
THREAD_SELF could be used for this, since it should get destroyed at the end of the thread's life.

Reply:
I am pretty sure it is totally unsafe to use a detached thread because it may free at any time. We could add some error handling for it.

Reply:
Actually that's true for any thread... It might not be feasible to store the data in a thread local since the thread could die at any time. I think the btreemap might be the best option then.

Reply:
I have some local changes which attempt to implement the `#[thread_local]` approach, which adds a flag for detached threads to free themselves on cleanup, and non-detached threads are freed by a call to `join` (or leaked, if they are never joined). 

From reading e.g. https://linux.die.net/man/3/pthread_detach, it seems like it's fair to say that it's invalid to attempt to join a previously detached thread, but I'm not sure what the expected behavior is (unspecified? return an error?). Similarly, I'm not sure what the behavior for attempting to make calls like `setschedparam` and `getschedparam` is intended to be...

FWIW, the only use of `pthread_detach` I can find in `std` is the `Drop` impl for `sys::unix::Thread`, which I believe means that no other threads can perform further operations that would dereference the `*Pthread` (but of course the thread could do so on itself).

>  It might not be feasible to store the data in a thread local since the thread could die at any time. I think the btreemap might be the best option then.

This is true, but the thread-local variable is only storing a _pointer_ to the thread data, which itself is on the heap – so the data remains valid until either another thread with a `JoinHandle` frees it, or it is detached and frees itself. I suppose there is a possibility of double-free if multiple threads attempt to join it, but I don't think it's allowed to copy join handles around anyway...

Reply:
Once the thread dies, the thread local state gets freed. So if you read from it, you're reading from a dangling pointer. Since a thread could die at any time, there's no safe way to use thread local state as the main data store.

Reply:
Right, but the thread-local is not actually holding the `PThread` struct, just a pointer to it on the heap. Either a joining thread can free it from heap (possibly well after the thread itself has exited), or a detached thread can free the `PThread` itself upon exit. I believe this is safe the way it is used in `std`, but perhaps I can verify by copying the tests from https://github.com/Meziu/rust-horizon/blob/horizon-std/library/std/src/thread/tests.rs into the `ctru-rs` test runner to verify – does that seem fair?

I have been traveling so haven't had a chance to test / push the changes I have locally, but will try to do so tonight – maybe it will make more sense to just show the code I have.

Reply:
> From reading e.g. https://linux.die.net/man/3/pthread_detach, it seems like it's fair to say that it's invalid to attempt to join a previously detached thread, but I'm not sure what the expected behavior is (unspecified? return an error?).

From this I think it's just EINVAL:
> EINVAL
    thread is not a joinable thread.
> ...
> Failure to join with a thread that is joinable (i.e., one that is not detached), produces a "zombie thread".

https://linux.die.net/man/3/pthread_join

Reply:
> Right, but the thread-local is not actually holding the `PThread` struct, just a pointer to it on the heap. Either a joining thread can free it from heap (possibly well after the thread itself has exited), or a detached thread can free the `PThread` itself upon exit.

Since we need to clean up the heap data when the thread dies, any future usage of the pthread_t value would be using a dangling pointer. If you're given a `pthread_t`, how do you know if the allocation it points to is still around (if the thread has finished)? I don't think there's enough data there to answer the question, but it is still valid to use a `pthread_t` after the thread has finished. For example, you can join on a finished thread (ex. thread may have already finished by the time the main thread calls join).

Reply:
By the way, the libctru implementation isn't unsafe even though it uses a similar heap allocation idea. It has a `threadFree` function which explicitly closes out the thread handle, which pthread doesn't have.